ffi: Expose discovered sliding sync proxy URL

vollkorntomate opened the pull request:
Closes #3265.

There is currently no way to get the URL of a homeserver's sliding sync proxy before logging in on the homeserver.

I suggest exposing the URL via the `HomeserverLoginDetail` struct after configuring the homeserver (through `configure_homeserver`).

Since the homeserver may not declare a corresponding sliding sync proxy, this value is an `Optional`. It is used later in `configure_homeserver` to check if a sliding sync proxy exists and throws an error otherwise. Previously, this check was done against the client's `discovered_sliding_sync_proxy` function.

- [ ] Public API changes documented in changelogs (optional)

---
Signed-off-by: Thomas Völkl <thomas@vollkorntomate.de>

bnjbvr reviewed (changes requested):
Thanks. That's a preexisting issue, but `discovered_sliding_sync_proxy` is a misnomer in the (non FFI) `Client`: if it was set with `set_sliding_sync_proxy` after discovery, then it's not the one discovered automatically. So it's also incorrect in this PR, and it should just be called `sliding_sync_proxy`. Could you tweak the name here? Bonus points if you tweak it in the main crate too, expand the doc comment of the field and add a changelog entry (since this is technically a breaking change of the public API of `Client`).

bnjbvr commented:
(also please include a signoff in your PR description, as requested by our contribution guidelines)

vollkorntomate commented:
Thanks for your comment, Benjamin!
I removed the `discovered_` part from the name as you suggested.
Could you please clarify what you mean with changing the name in the main crate as well? As I see it, the corresponding field and functions in `matrix-sdk` crate' `Client` are already named `sliding_sync_proxy` without any "discovered".

Also, I added the sign-off to the PR description.

bnjbvr commented on bindings/matrix-sdk-ffi/src/authentication_service.rs:
```diff
@@ -198,6 +199,11 @@ impl HomeserverLoginDetails {
         self.url.clone()
     }
 
+    /// The URL of the discovered sliding sync proxy, if any
```
```suggestion
    /// The URL of the discovered or manually set sliding sync proxy, if any
```

bnjbvr reviewed (approved):
Ah you're right, I was looking at the wrong `Client`, it's only the one in FFI that was impacted.

Thanks for the patch!

bnjbvr commented:
In addition to my comment above: Can you run `cargo xtask style fixup` please, so it fixes rustfmt? Thanks!

bnjbvr commented:
Sorry, you're hitting intermittent failures in the code coverage task that have started to become more, uh, permanent. I'm investigating the failure separately.
